retrieval/discovery/kubernetes.go

```diff
@@ -0,0 +1,593 @@
+				log.Errorf("Unable to watch services: %s", err)
+				return
+			}
+			kd.servicesResourceVersion = event.Service.ObjectMeta.ResourceVersion
+			events <- &event
+		}
+	}, retryDelay, kd.runDone)
+}
+
+func (kd *KubernetesDiscovery) updateService(service *kubernetes.Service, eventType kubernetes.EventType, ch chan<- *config.TargetGroup) {
+	kd.servicesMu.Lock()
+	defer kd.servicesMu.Unlock()
+	name := service.ObjectMeta.Name
+	namespace := service.ObjectMeta.Namespace
+	scrape := kd.checkScrape(service)
+	_, ok := kd.services[namespace][name]
+	switch eventType {
+	case kubernetes.Deleted:
+		if ok {
+			kd.deleteService(service, ch)
+		}
+	case kubernetes.Modified:
+		// Exists & no scrape - delete.
+		if ok && !scrape {
+			kd.deleteService(service, ch)
+			return
+		}
+
+		if scrape {
+			// Must be scrapable so let's add it to ensure labels are updated properly at least.
+			kd.addService(service, ch)
+		}
+	case kubernetes.Added:
+		if scrape {
+			kd.addService(service, ch)
+		}
+	}
+}
+
+func (kd *KubernetesDiscovery) deleteService(service *kubernetes.Service, ch chan<- *config.TargetGroup) {
+	tg := &config.TargetGroup{Source: kubernetesSourceServicePrefix + ":" + service.ObjectMeta.Namespace + "/" + service.ObjectMeta.Name}
+	ch <- tg
+	delete(kd.services[service.ObjectMeta.Namespace], service.ObjectMeta.Name)
+	if len(kd.services[service.ObjectMeta.Namespace]) == 0 {
+		delete(kd.services, service.ObjectMeta.Namespace)
+	}
+}
+
+func (kd *KubernetesDiscovery) addService(service *kubernetes.Service, ch chan<- *config.TargetGroup) {
+	namespace, ok := kd.services[service.ObjectMeta.Namespace]
+	if !ok {
+		namespace = make(map[string]*kubernetes.Service)
+		kd.services[service.ObjectMeta.Namespace] = namespace
+	}
+	namespace[service.ObjectMeta.Name] = service
+	endpointURL := fmt.Sprintf(kubernetesServiceEndpointsURL, service.ObjectMeta.Namespace, service.ObjectMeta.Name)
+	res, err := kd.client.Get(kd.conf.Server + endpointURL)
+	if err != nil {
+		log.Errorf("Error getting service endpoints: %s", err)
+		return
+	}
+	if res.StatusCode != http.StatusOK {
+		log.Errorf("Failed to get service endpoints: %s", res.StatusCode)
+		return
+	}
+
+	var endpoints kubernetes.Endpoints
+	if err := json.NewDecoder(res.Body).Decode(&endpoints); err != nil {
+		log.Errorf("Error getting service endpoints: %s", err)
+		return
+	}
+
+	kd.updateServiceTargetGroup(service, &endpoints, ch)
+}
+
+func (kd *KubernetesDiscovery) updateServiceTargetGroup(service *kubernetes.Service, endpoints *kubernetes.Endpoints, ch chan<- *config.TargetGroup) {
+	tg := &config.TargetGroup{
+		Source: kubernetesSourceServicePrefix + ":" + service.ObjectMeta.Namespace + "/" + service.ObjectMeta.Name,
+		Labels: clientmodel.LabelSet{
+			kubernetesServiceNamespaceLabel: clientmodel.LabelValue(service.ObjectMeta.Namespace),
+			kubernetesServiceNameLabel:      clientmodel.LabelValue(service.ObjectMeta.Name),
+		},
+	}
+
+	for k, v := range service.ObjectMeta.Labels {
```

Comment:
Thinking a bit, there's also a problem here as Kubernetes allows changing of labels after an object is created. Target labels should be constant over the life of the target to avoid all the timeseries changing names, as that leads to discontinuities in graphs and can cause other problems.